pkg/extensions/sync/on_demand.go

```diff
@@ -3,29 +3,51 @@ package sync
 import (
 	"context"
 	"fmt"
+	"net/url"
 	"os"
-	"path"
+	"sync"
+	"time"
 
 	"github.com/containers/common/pkg/retry"
 	"github.com/containers/image/v5/copy"
-	"github.com/containers/image/v5/docker"
-	"github.com/containers/image/v5/docker/reference"
-	"github.com/containers/image/v5/oci/layout"
-	guuid "github.com/gofrs/uuid"
+	"gopkg.in/resty.v1"
 	"zotregistry.io/zot/pkg/log"
 	"zotregistry.io/zot/pkg/storage"
 )
 
+// nolint: gochecknoglobals
+var demandedImgs demandedImages
+
+type demandedImages struct {
+	sync.Map
+}
+
 func OneImage(cfg Config, storeController storage.StoreController,
-	repo, tag string, log log.Logger) error {
-	var credentialsFile CredentialsFile
+	repo, tag string, isArtifact bool, log log.Logger) error {
+	// guard against multiple parallel requests
+	demandedImage := fmt.Sprintf("%s:%s", repo, tag)
+	// loadOrStore image-based channel
+	value, found := demandedImgs.LoadOrStore(demandedImage, make(chan error))
+	imageChannel, _ := value.(chan error)
```

Comment:
why typecast, just return a proper type from LoadOrStore, or change the API signatures, no multipurpose stuff pls.

Reply:
this is how sync.Map works :d, I made a wrapper around it.